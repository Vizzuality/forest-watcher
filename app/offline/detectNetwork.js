import detectNetwork from '@redux-offline/redux-offline/lib/defaults/detectNetwork.native';
import Config from 'react-native-config';

import checkConnectivity from 'helpers/networking';

export class DetectNetworkPing {
  static urlList = [
    Config.API_URL,
    'https://www.globalforestwatch.org',
    'https://www.google.com',
    'https://www.facebook.com',
    'https://www.amazon.com'
  ];

  pingToDetectNetwork = (dispatch, urlIndex) => connection => {
    /*
      This is being called when the connection status is updated (like when the WiFi / LTE is being enabled / disabled).
      If everything is off, then the connection would be offline so thereâ€™d be no need to continue.
      When it detects that WiFi or LTE has been enabled again, it should become true and then we can check for internet reachability.
      Redux-offline should handle all of that internally.
    */
    if (!connection.online) {
      dispatch(connection);
      return;
    }

    // Get URL based on current attempt number.
    const url = DetectNetworkPing.urlList[urlIndex % DetectNetworkPing.urlList.length];

    // Attempt fetch with a 1s timeout.
    return checkConnectivity(url).then(connected => {
      // If we've got a connection, update the redux state.
      if (connected) {
        dispatch({ ...connection, online: true });
        return;
      }

      // There was an error, so at this point we should state the device is offline.
      if (urlIndex === 0) {
        dispatch({ ...connection, online: false });
      }

      if (urlIndex < DetectNetworkPing.urlList.length - 1) {
        // Recall this function, incrementing the urlIndex so we try the next URL.
        this.pingToDetectNetwork(dispatch, urlIndex + 1)(connection);
      } else {
        // If every URL has failed, fail the request and stop attempting.
        dispatch({ ...connection, online: false });
      }
    });
  };

  /**
   *  Starts the network detection, with the starting URL index of 0.
   */
  start = dispatch => detectNetwork(this.pingToDetectNetwork(dispatch, 0));
}

const detector = new DetectNetworkPing();

export default detector.start;
