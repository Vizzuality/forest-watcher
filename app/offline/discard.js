/**
 * Modification of the default redux-offline discard implementation, that will also discard after server 500+ errors.
 *
 * @param error
 * @param action
 * @param retries
 * @return {boolean}  True if the action should be discarded, false otherwise
 */
export default (error, action, retries) => {
  const status = error?.status || 0;
  return status >= 400;
};
