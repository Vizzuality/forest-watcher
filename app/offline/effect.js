import defaultEffect from 'redux-offline/lib/defaults/effect';

const JSONAPIDeserializer = require('jsonapi-serializer').Deserializer;

const deserializeOptions = {
  keyForAttribute: 'camelCase'
};

export default function effect({ url, headers, promise, errorCode, deserialize = true, ...params }, { auth, ...action }) {
  if (url && typeof url === 'string') {
    const req = {
      ...params,
      url,
      headers: { ...headers, Authorization: `Bearer ${auth}` }
    };
    const canDeserialize = ({ data }) => (typeof data === 'object' && data && deserialize);
    return defaultEffect(req, action)
      .then((data) => (canDeserialize(data) ? new JSONAPIDeserializer(deserializeOptions).deserialize(data) : data));
  } else if (typeof promise !== 'undefined') {
    return promise
      .then(data => data)
      .catch(err => {
        console.warn(err);
        return Promise.reject(errorCode);
      });
  }
  throw new TypeError();
}

