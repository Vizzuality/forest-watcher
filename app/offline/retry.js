const decaySchedule = [
  1000 * 5, // After 5 seconds
  1000 * 15, // After 20 seconds
  1000 * 60 // After 1 minute
];

export default (action, retries) => decaySchedule[retries] || null;
