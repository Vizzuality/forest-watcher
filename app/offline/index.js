import { offline } from '@redux-offline/redux-offline';
import offlineConfig from '@redux-offline/redux-offline/lib/defaults';
import detectNetwork from '@redux-offline/redux-offline/lib/defaults/detectNetwork.native';
import { AsyncStorage } from 'react-native'; // eslint-disable-line import/no-unresolved
import { persistStore } from 'redux-persist';
import { version } from 'package.json';
import { resetAlertsDb } from 'redux-modules/alerts';
import effect from './effect';
import retry from './retry';

const persistNative = (store, options, callback) => {
  AsyncStorage.getItem('reduxPersist:app', (err, appData) => {
    const getPersistedStore = () => persistStore(store, { storage: AsyncStorage, ...options }, callback);
    let app = null;
    if (!err) {
      app = JSON.parse(appData);
    }
    if (app && app.version !== version) {
      getPersistedStore().purge();
      resetAlertsDb();
    } else {
      getPersistedStore(); // .purge to clean the offline data
    }
  });
};

const config = params => ({
  ...offlineConfig,
  effect,
  retry,
  detectNetwork,
  persist: persistNative,
  persistOptions: { blacklist: ['setup'] },
  persistCallback: params.persistCallback
});

export default params => offline(config(params));
