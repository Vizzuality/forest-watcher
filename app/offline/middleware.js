import { OFFLINE_STATUS_CHANGED } from 'redux-offline/lib/constants';
import { NET_INFO_CHANGED } from 'redux-modules/app';

// I'm not very fond of this solution. We should PR redux-offline in the future and handle this inside it
export const netInfoMiddleware = ({ dispatch }) => next => action => {
  if (action.type === OFFLINE_STATUS_CHANGED) {
    const { netInfo, online } = action.payload.online || {};
    dispatch({ type: NET_INFO_CHANGED, payload: netInfo });
    return next({ type: action.type, payload: { online } });
  }
  return next(action);
};

export default { netInfoMiddleware };
