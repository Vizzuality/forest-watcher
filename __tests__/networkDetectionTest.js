import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import Config from 'react-native-config';
import checkConnectivity from 'helpers/networking';
import { DetectNetworkPing } from 'offline/detectNetwork';

global.fetch = require('jest-fetch-mock');

const mockStore = configureMockStore([thunk]);

const STATUS_CHANGED = 'STATUS_CHANGED';

describe('detect network actions', () => {
  afterEach(() => {
    fetch.resetMocks();
  });

  it('connectivity request when rejected returns false', () => {
    fetch.mockRejectOnce({ ok: false });

    checkConnectivity(Config.API_URL).then(response => {
      expect(response).toEqual(false);
    });
  });

  it('connectivity request when connected returns true', () => {
    fetch.mockResponseOnce(JSON.stringify({ ok: true }));

    checkConnectivity(Config.API_URL).then(response => {
      expect(response).toEqual(true);
    });
  });

  it('pingToDetectNetwork when first url is alive when online sends online action', async () => {
    const store = mockStore();

    fetch.mockResponseOnce(JSON.stringify({ ok: true }));

    const networkDetector = new DetectNetworkPing();

    await networkDetector.pingToDetectNetwork(store.dispatch, 0)({ type: STATUS_CHANGED, online: true });

    const expectedActions = [{ type: STATUS_CHANGED, online: true }];

    expect(store.getActions()).toEqual(expectedActions);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('pingToDetectNetwork when first url is alive when offline sends offline action', async () => {
    const store = mockStore();

    fetch.mockResponseOnce(JSON.stringify({ ok: true }));

    const networkDetector = new DetectNetworkPing();

    await networkDetector.pingToDetectNetwork(store.dispatch, 0)({ type: STATUS_CHANGED, online: false });

    const expectedActions = [{ type: STATUS_CHANGED, online: false }];

    expect(store.getActions()).toEqual(expectedActions);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('pingToDetectNetwork when first url is dead and second is online sends offline and online actions', async () => {
    const store = mockStore();

    fetch.mockResponses([JSON.stringify({ ok: false }), JSON.stringify({ ok: true })]);

    const networkDetector = new DetectNetworkPing();

    await networkDetector.pingToDetectNetwork(store.dispatch, 0)({ type: STATUS_CHANGED, online: true });

    const expectedActions = [{ type: STATUS_CHANGED, online: true }];

    expect(store.getActions()).toEqual(expectedActions);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('pingToDetectNetwork when all urls are dead sends offline action', async () => {
    const store = mockStore();

    fetch.mockReject(JSON.stringify({ ok: false }));

    const networkDetector = new DetectNetworkPing();

    await networkDetector.pingToDetectNetwork(store.dispatch, 0)({ type: STATUS_CHANGED, online: true });

    const expectedActions = [{ type: STATUS_CHANGED, online: false }];

    expect(store.getActions()).toEqual(expectedActions);
    expect(store.getActions()).toMatchSnapshot();
  });
});
