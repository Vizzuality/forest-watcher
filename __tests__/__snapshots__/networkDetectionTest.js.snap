// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`detect network actions pingToDetectNetwork when all urls are dead sends offline action 1`] = `
Array [
  Object {
    "online": false,
    "type": "STATUS_CHANGED",
  },
]
`;

exports[`detect network actions pingToDetectNetwork when first url is alive when offline sends offline action 1`] = `
Array [
  Object {
    "online": false,
    "type": "STATUS_CHANGED",
  },
]
`;

exports[`detect network actions pingToDetectNetwork when first url is alive when online sends online action 1`] = `
Array [
  Object {
    "online": true,
    "type": "STATUS_CHANGED",
  },
]
`;

exports[`detect network actions pingToDetectNetwork when first url is dead and second is online sends offline and online actions 1`] = `
Array [
  Object {
    "online": true,
    "type": "STATUS_CHANGED",
  },
]
`;
